ps3.py

```diff
@@ -0,0 +1,401 @@
+            if hand_update[i] < 0:
+                hand_update[i] = 0
+        if i not in hand:
+            continue
+    return hand_update
+
+
+#
+# Problem #3: Test word validity
+#
+def is_valid_word(word, hand, word_list):
+    """
+    Returns True if word is in the word_list and is entirely
+    composed of letters in the hand. Otherwise, returns False.
+    Does not mutate hand or word_list.
+   
+    word: string
+    hand: dictionary (string -> int)
+    word_list: list of lowercase strings
+    returns: boolean
+    """
+
+    hand_lst = list("".join([k * v for k, v in hand.copy().items()]))
+    for letter in word.lower():
+        if letter in hand_lst:
+            hand_lst.remove(letter)
+        else:
+            return False
+
+    if WILDCARD in word.lower():
+        for vowels in VOWELS:
+            if word.replace(WILDCARD, vowels) in word_list:
+                return True
+        return False
+    elif word.lower() in word_list:
+        return True
+    return False
+
+
+#
+# Problem #5: Playing a hand
+#
+def calculate_handlen(hand):
+    """ 
+    Returns the length (number of letters) in the current hand.
+    
+    hand: dictionary (string-> int)
+    returns: integer
+    """
+    count_letter = 0
+    for letter in hand.keys():
+        count_letter += hand[letter]
+    return count_letter
+
+
+def play_hand(hand, word_list):
+    """
+    Allows the user to play the given hand, as follows:
+
+    * The hand is displayed.
+    
+    * The user may input a word.
+
+    * When any word is entered (valid or invalid), it uses up letters
+      from the hand.
+
+    * An invalid word is rejected, and a message is displayed asking
+      the user to choose another word.
+
+    * After every valid word: the score for that word is displayed,
+      the remaining letters in the hand are displayed, and the user
+      is asked to input another word.
+
+    * The sum of the word scores is displayed when the hand finishes.
+
+    * The hand finishes when there are no more unused letters.
+      The user can also finish playing the hand by inputing two 
+      exclamation points (the string '!!') instead of a word.
+
+      hand: dictionary (string -> int)
+      word_list: list of lowercase strings
+      returns: the total score for the hand
+      
+    """
+    score = 0
+    while hand != {}:
+        print('\nCurrent hand: ', end='')
+        display_hand(hand)
+        word = input('Enter word, or “!!” to indicate that you are finished: ')
+        if word == '!!':
+            break
+        elif is_valid_word(word, hand, word_list):
+            score += get_word_score(word, HAND_SIZE)
+            print('"' + word + '" earned', get_word_score(word, HAND_SIZE), 'points. Total:', score, 'points')
+        else:
+            print('This is not a valid word.', end='')
+            if update_hand(hand, word) != {}:
+                print(' Please choose another word.')
+            else:
+                print()
```

Comment:
I think it's necessary part of code. You can combine this print with a print above.